crates/chain/src/tx_graph.rs

```diff
@@ -52,28 +54,82 @@
 //! let additions = graph.apply_update(update);
 //! assert!(additions.is_empty());
 //! ```
-use crate::{collections::*, ForEachTxOut};
+
+use crate::{collections::*, BlockAnchor, BlockId, ChainOracle, ForEachTxOut, ObservedIn};
 use alloc::vec::Vec;
 use bitcoin::{OutPoint, Transaction, TxOut, Txid};
-use core::ops::RangeInclusive;
+use core::{
+    convert::Infallible,
+    ops::{Deref, RangeInclusive},
+};
 
 /// A graph of transactions and spends.
 ///
 /// See the [module-level documentation] for more.
 ///
 /// [module-level documentation]: crate::tx_graph
-#[derive(Clone, Debug, PartialEq, Default)]
-pub struct TxGraph {
-    txs: HashMap<Txid, TxNode>,
+#[derive(Clone, Debug, PartialEq)]
+pub struct TxGraph<A = BlockId> {
+    // all transactions that the graph is aware of in format: `(tx_node, tx_anchors, tx_last_seen)`
+    txs: HashMap<Txid, (TxNode, BTreeSet<A>, u64)>,
     spends: BTreeMap<OutPoint, HashSet<Txid>>,
+    anchors: BTreeSet<(A, Txid)>,
 
     // This atrocity exists so that `TxGraph::outspends()` can return a reference.
     // FIXME: This can be removed once `HashSet::new` is a const fn.
     empty_outspends: HashSet<Txid>,
 }
 
-/// Node of a [`TxGraph`]. This can either be a whole transaction, or a partial transaction (where
-/// we only have select outputs).
+impl<A> Default for TxGraph<A> {
+    fn default() -> Self {
+        Self {
+            txs: Default::default(),
+            spends: Default::default(),
+            anchors: Default::default(),
+            empty_outspends: Default::default(),
+        }
+    }
+}
+
+// pub type InChainTx<'a, T, A> = (ObservedIn<&'a A>, TxInGraph<'a, T, A>);
+// pub type InChainTxOut<'a, I, A> = (&'a I, FullTxOut<ObservedIn<&'a A>>);
+
+/// An outward-facing view of a transaction that resides in a [`TxGraph`].
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct TxInGraph<'a, T, A> {
+    /// Txid of the transaction.
+    pub txid: Txid,
+    /// A partial or full representation of the transaction.
+    pub tx: &'a T,
+    /// The blocks that the transaction is "anchored" in.
+    pub anchors: &'a BTreeSet<A>,
+    /// The last-seen unix timestamp of the transaction.
+    pub last_seen: u64,
```

Comment:
Call this `last_seen_unconfirmed`. The doc comment needs to properly explain what this means.

Reply:
Fixed in 7810059ed0f23cae7dee61fe587a1c8f3f49480a